File: Data Structures & Algorithms/eating-bananas/submission-0.go
/*
 * @lc app=leetcode id=875 lang=golang
 *
 * [875] Koko Eating Bananas
 */


import (
	"slices"
)

// @lc code=start
func minEatingSpeed(piles []int, h int) int {
	if len(piles) == h {
        return slices.Max(piles) 
    }

    left, right := 1, slices.Max(piles)
	result := 1
	
	for left <= right {
		mid := left + (right - left) / 2

		if canEatAll(piles, mid, h) {
			right = mid - 1
			result = mid
		} else {
			left = mid + 1
		}
	}

	return result
}

func canEatAll(piles []int, k int, h int) bool {
    total := 0

    for _, p := range piles {
        total += (p + k - 1) / k

        if total > h {
            return false
        }
    }

    return total <= h
}
// @lc code=end

File: Data Structures & Algorithms/median-of-two-sorted-arrays/submission-0.cpp
/*
 * @lc app=leetcode id=4 lang=cpp
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start

#include <vector>

class Solution {
public:
    double findMedianSortedArrays(std::vector<int>& nums1, std::vector<int>& nums2) {
        if (nums1.size() > nums2.size())
        {
            return findMedianSortedArrays(nums2, nums1);
        }

        if (nums1.empty())
        {
            return nums2.size() % 2 == 0 ? (nums2[nums2.size() / 2] + nums2[nums2.size() / 2 - 1]) / 2.0 : nums2[nums2.size() / 2];
        }

        int left = 0, right = nums1.size();
        int max_left, min_right;

        auto total_left = (nums1.size() + nums2.size() + 1) / 2;

        while (left <= right) {
            auto mid1 = left + (right - left) / 2;
            auto mid2 = total_left - mid1;

            if (mid1 > 0 && nums1[mid1 - 1] > nums2[mid2])
            {
                right = mid1 - 1;
                continue;
            }

            if (mid1 != nums1.size() && nums1[mid1] < nums2[mid2 - 1])
            {
                left = mid1 + 1;
                continue;
            }
            
            if (mid1 == 0)
            {
                max_left = nums2[mid2 - 1];
            }
            else if (mid2 == 0)
            {
                max_left = nums1[mid1 - 1];
            }
            else
            {
                max_left = std::max(nums1[mid1 - 1], nums2[mid2 - 1]);
            }

            if (mid1 == nums1.size())
            {
                min_right = nums2[mid2];
            }
            else if (mid2 == nums2.size())
            {
                min_right = nums1[mid1];
            }
            else
            {
                min_right = std::min(nums1[mid1], nums2[mid2]);
            }

            if ((nums1.size() + nums2.size()) % 2 == 0)
            {
                return (max_left + min_right) / 2.0;
            }
            else
            {
                return max_left;
            }
        }

        return 0.0;
    }
};
// @lc code=end
